### PR TITLE
kata-manager: Fix deployment of containerd on architectures other than amd64.

### PR DESCRIPTION
**Pull request by brianwang12**
This patch could fix the error encountered when attempting to deploy containerd on an arm machine(https://github.com/kata-containers/kata-containers/issues/7071), 
the error message : 
_ERROR: Cannot determine download URL for version v1.7.2 (https://api.github.com/repos/containerd/containerd/releases)_
This patch has been verified on both arm and x86 machines, the logs are as followed:

```
./kata-manager.sh  -k 3.2.0-alpha3 -c v1.7.2
INFO: Checking dependencies
INFO: Running pre-checks
INFO: Downloading Kata Containers release (version 3.2.0-alpha3)
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
100  130M  100  130M    0     0  12.9M      0  0:00:10  0:00:10 --:--:-- 18.9M
INFO: Installing Kata Containers release 3.2.0-alpha3 from /tmp/tmp.K5HjMajCuO/kata-static-3.2.0-alpha3-aarch64.tar.xz
[sudo] password for briwan01: 
./
./opt/
./opt/kata/
./opt/kata/bin/
./opt/kata/bin/containerd-shim-kata-v2
./opt/kata/bin/firecracker
./opt/kata/bin/kata-monitor
./opt/kata/bin/jailer
./opt/kata/bin/qemu-system-aarch64
./opt/kata/bin/kata-collect-data.sh
./opt/kata/bin/kata-runtime
./opt/kata/bin/cloud-hypervisor
./opt/kata/libexec/
./opt/kata/libexec/nydusd
./opt/kata/libexec/virtiofsd
./opt/kata/share/
./opt/kata/share/kata-qemu/
./opt/kata/share/kata-qemu/qemu/
./opt/kata/share/kata-qemu/qemu/edk2-i386-code.fd
./opt/kata/share/kata-qemu/qemu/qboot.rom
./opt/kata/share/kata-qemu/qemu/hppa-firmware.img
./opt/kata/share/kata-qemu/qemu/s390-netboot.img
./opt/kata/share/kata-qemu/qemu/vof-nvram.bin
./opt/kata/share/kata-qemu/qemu/s390-ccw.img
./opt/kata/share/kata-qemu/qemu/bios.bin
./opt/kata/share/kata-qemu/qemu/edk2-arm-code.fd
./opt/kata/share/kata-qemu/qemu/edk2-x86_64-code.fd
./opt/kata/share/kata-qemu/qemu/bios-256k.bin
./opt/kata/share/kata-qemu/qemu/multiboot_dma.bin
./opt/kata/share/kata-qemu/qemu/vof.bin
./opt/kata/share/kata-qemu/qemu/qemu-nsis.bmp
./opt/kata/share/kata-qemu/qemu/bios-microvm.bin
./opt/kata/share/kata-qemu/qemu/linuxboot.bin
./opt/kata/share/kata-qemu/qemu/linuxboot_dma.bin
./opt/kata/share/kata-qemu/qemu/pvh.bin
./opt/kata/share/kata-qemu/qemu/firmware/
./opt/kata/share/kata-qemu/qemu/firmware/50-edk2-i386-secure.json
./opt/kata/share/kata-qemu/qemu/firmware/60-edk2-i386.json
./opt/kata/share/kata-qemu/qemu/firmware/60-edk2-aarch64.json
./opt/kata/share/kata-qemu/qemu/firmware/60-edk2-arm.json
./opt/kata/share/kata-qemu/qemu/firmware/50-edk2-x86_64-secure.json
./opt/kata/share/kata-qemu/qemu/firmware/60-edk2-x86_64.json
./opt/kata/share/kata-qemu/qemu/kvmvapic.bin
./opt/kata/share/kata-qemu/qemu/efi-virtio.rom
./opt/kata/share/kata-qemu/qemu/edk2-licenses.txt
./opt/kata/share/kata-qemu/qemu/edk2-i386-secure-code.fd
./opt/kata/share/kata-qemu/qemu/edk2-aarch64-code.fd
./opt/kata/share/kata-qemu/qemu/edk2-i386-vars.fd
./opt/kata/share/kata-qemu/qemu/edk2-x86_64-secure-code.fd
./opt/kata/share/kata-qemu/qemu/edk2-arm-vars.fd
./opt/kata/share/bash-completion/
./opt/kata/share/bash-completion/completions/
./opt/kata/share/bash-completion/completions/kata-runtime
./opt/kata/share/defaults/
./opt/kata/share/defaults/kata-containers/
./opt/kata/share/defaults/kata-containers/configuration-clh.toml
./opt/kata/share/defaults/kata-containers/configuration-dragonball.toml
./opt/kata/share/defaults/kata-containers/configuration.toml
./opt/kata/share/defaults/kata-containers/configuration-qemu-snp.toml
./opt/kata/share/defaults/kata-containers/configuration-qemu-tdx.toml
./opt/kata/share/defaults/kata-containers/configuration-qemu-sev.toml
./opt/kata/share/defaults/kata-containers/configuration-qemu.toml
./opt/kata/share/defaults/kata-containers/configuration-qemu-nvidia-gpu.toml
./opt/kata/share/defaults/kata-containers/configuration-fc.toml
./opt/kata/share/kata-containers/
./opt/kata/share/kata-containers/vmlinuz-dragonball-experimental.container
./opt/kata/share/kata-containers/vmlinuz-5.19.2-105
./opt/kata/share/kata-containers/vmlinux-5.10.25-105-dragonball-experimental
./opt/kata/share/kata-containers/kata-containers.img
./opt/kata/share/kata-containers/kata-containers-initrd.img
./opt/kata/share/kata-containers/vmlinuz.container
./opt/kata/share/kata-containers/vmlinux.container
./opt/kata/share/kata-containers/vmlinux-dragonball-experimental.container
./opt/kata/share/kata-containers/config-5.19.2
./opt/kata/share/kata-containers/kata-ubuntu-latest.image
./opt/kata/share/kata-containers/config-5.10.25
./opt/kata/share/kata-containers/vmlinuz-5.10.25-105-dragonball-experimental
./opt/kata/share/kata-containers/kata-alpine-3.15.initrd
./opt/kata/share/kata-containers/vmlinux-5.19.2-105
./opt/kata/runtime-rs/
./opt/kata/runtime-rs/bin/
./opt/kata/runtime-rs/bin/containerd-shim-kata-v2
INFO: Kata Containers installed

INFO: Using default Kata Containers configuration
kata-runtime  : 3.2.0-alpha3
   commit   : 9aae3333437948764324f6e51e126fce8808c26b
   OCI specs: 1.0.2-dev
INFO: Downloading containerd release (version v1.7.2)
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
100 33.0M  100 33.0M    0     0  8354k      0  0:00:04  0:00:04 --:--:-- 13.3M
INFO: Installing containerd release v1.7.2 from /tmp/tmp.K5HjMajCuO/containerd-1.7.2-linux-arm64.tar.gz
bin/
bin/containerd-shim-runc-v1
bin/containerd-shim-runc-v2
bin/containerd-stress
bin/containerd
bin/containerd-shim
bin/ctr
INFO: containerd installed

INFO: Configuring containerd
INFO: Configured containerd (debug disabled)

containerd github.com/containerd/containerd v1.7.2 0cae528dd6cb557f7201036e9f43420650207b58
INFO: Testing Kata Containers

INFO[0000] Setting 'disable_image_nvdimm = true' if there is no firmware specified 
INFO[0000] IOMMUPlatform is disabled by default.        
WARN[0000] Not running network checks as super user      arch=arm64 name=kata-runtime pid=14501 source=runtime
INFO[0000] Unable to know if the system is running inside a VM  arch=arm64 source=virtcontainers/hypervisor
INFO[0000] kernel property found                         arch=arm64 description="Kernel-based Virtual Machine" name=kvm pid=14501 source=runtime type=module
INFO[0000] kernel property found                         arch=arm64 description="Host kernel accelerator for virtio" name=vhost pid=14501 source=runtime type=module
INFO[0000] kernel property found                         arch=arm64 description="Host kernel accelerator for virtio network" name=vhost_net pid=14501 source=runtime type=module
INFO[0000] kernel property found                         arch=arm64 description="Host Support for Linux VM Sockets" name=vhost_vsock pid=14501 source=runtime type=module
System is capable of running Kata Containers
INFO[0000] device available                              arch=arm64 check-type=full device=/dev/kvm name=kata-runtime pid=14501 source=runtime
INFO[0000] feature available                             arch=arm64 check-type=full feature=create-vm name=kata-runtime pid=14501 source=runtime
INFO[0000] kvm extension is supported                    arch=arm64 description="Maximum IPA shift supported by the host" id=165 name=KVM_CAP_ARM_VM_IPA_SIZE pid=14501 source=runtime type="kvm extension"
INFO[0000] IPA limit size: 48 bits.                      arch=arm64 name=KVM_CAP_ARM_VM_IPA_SIZE pid=14501 source=runtime type="kvm extension"
System can currently create Kata Containers
docker.io/library/busybox:latest:                                                 resolved       |++++++++++++++++++++++++++++++++++++++| 
index-sha256:560af6915bfc8d7630e50e212e08242d37b63bd5c1ccf9bd4acccf116e262d5b:    exists         |++++++++++++++++++++++++++++++++++++++| 
manifest-sha256:3be7f0919fc88e394cf86dc712cfb747f4ea29bd3e4e75508c3f53aeb2135ad7: done           |++++++++++++++++++++++++++++++++++++++| 
config-sha256:f92f3ea6e4a8c34d0521435aeae960a08746e2dc17ce0b75c0f7603fe0b9dc6a:   done           |++++++++++++++++++++++++++++++++++++++| 
layer-sha256:fee740108510eb41b69a19b84ac80616a932edeea714428428914774d373045e:    done           |++++++++++++++++++++++++++++++++++++++| 
elapsed: 2.5 s                                                                    total:   0.0 B (0.0 B/s)                                         
unpacking linux/arm64/v8 sha256:560af6915bfc8d7630e50e212e08242d37b63bd5c1ccf9bd4acccf116e262d5b...
done: 7.111368ms	
INFO: Test successful:

INFO:   Host kernel version      : 6.1.8+
INFO:   Container kernel version : 5.19.2

INFO: Kata Containers and containerd are now installed

WARNINGS:

- Use distro-packages where possible

  If your distribution packages Kata Containers, you should use these packages rather
  than running this script.

- Packages will **not** be automatically updated

  Since a package manager is not being used, it is **your** responsibility
  to ensure these packages are kept up-to-date when new versions are released
  to ensure you are using a version that includes the latest security and bug fixes.

- Potentially untested versions or version combinations

  This script installs the *newest* versions of Kata Containers
  and containerd from binary release packages. These versions may
  not have been tested with your distribution version.
```

### CODE REVIEW

**Comment by brianwang12** [comment]
@jongwu

**Comment by brianwang12** [comment]
> @brianwang12, thanks a lot for the patch.
> 
> This, for sure, will fix the problem with containerd. However, you'll face a problem downloading Kata Containers artefacts just after solving it.
> 
> I'd recommend to address that as part of the PR as well. Also, mind that we only release multi-arch artefacts from 3.2.0-alpha3 release.

Thanks for your quick reply, I appreciate your feedback and clarification regarding the issue of downloading Kata Containers artifacts. I apologize for any confusion caused by my previous message. I don't encounter "downloading Kata Containers artifacts "during deployment using kata-manager, the logs can be found above in my first comment

**Comment by fidencio** [comment]
> > @brianwang12, thanks a lot for the patch.
> > This, for sure, will fix the problem with containerd. However, you'll face a problem downloading Kata Containers artefacts just after solving it.
> > I'd recommend to address that as part of the PR as well. Also, mind that we only release multi-arch artefacts from 3.2.0-alpha3 release.
> 
> Thanks for your quick reply, I appreciate your feedback and clarification regarding the issue of downloading Kata Containers artifacts. I apologize for any confusion caused by my previous message. I don't encounter "downloading Kata Containers artifacts "during deployment using kata-manager, the logs can be found above in my first comment

Oh, I'm glad to know it works to get the correct aftefacts for the architecture you're using!

**Review by fidencio — APPROVED** [review]
lgtm, thanks @brianwang12!

**Review by GabyCT — APPROVED** [review]

**Review by cmaf — APPROVED** [review]

**Comment by cmaf** [comment]
Hi @brianwang12, could you make sure your commit is formatted like the example in the [contributing guide](https://github.com/kata-containers/community/blob/main/CONTRIBUTING.md#main-patch) so the tests pass? Otherwise LGTM

**Comment by brianwang12** [comment]
> Hi @brianwang12, could you make sure your commit is formatted like the example in the [contributing guide](https://github.com/kata-containers/community/blob/main/CONTRIBUTING.md#main-patch) so the tests pass? Otherwise LGTM

Thanks, I've formatted my patch following this guide.

**Comment by jongwu** [comment]
Thanks @brianwang12 -,  I test it on Arm, it works.

@BbolroC @Amulyam24-, mind help test on s390x and ppc64?

**Comment by BbolroC** [comment]
I've verified that this works on s390x. Thanks @brianwang12!

**Comment by Amulyam24** [comment]
Thanks for the fix @brianwang12 . I have not tested it because installing kata using `kata-manager` is not supported on Power yet(support for static tarball in progress) but the this fix should work fine on ppc64le too 👍🏼

**Comment by jongwu** [comment]
Hi @brianwang12 -, pls push again to trigger those "on-push" check.

File: utils/kata-manager.sh
Comment (jodh-intel) [inline utils/kata-manager.sh:152]: The code as-is is very clear, but it might be worth considering collapsing each case into a single line for quicker reading, For example:

```suggestion
				x86_64*) arch=amd64 ;;
```

File: utils/kata-manager.sh
Comment (jodh-intel) [inline utils/kata-manager.sh:163]: Nit: It should be "or" not "and":

```suggestion
				*)
					die "Unsupported arch. Must be x86_64, arm64, s390x, or ppc64le." ;;
```

File: utils/kata-manager.sh
Comment (jodh-intel) [inline utils/kata-manager.sh:153]: Nit: Unordered list :smile:

**Review by jodh-intel — APPROVED** [review]
Thanks @brianwang12.

A few nits, but overall lgtm.

Note that this PR conflicts with #8323. Since #8283 contains a containerd fix this PR will also need, I think it would make sense if we try to land #8283 first, then this one.

Please could the other architecture representatives test this PR on their respective arches (maybe once #8283 has landed and this has been rebased?)

/cc @justin-he, @BbolroC.

**Comment by jodh-intel** [comment]
Hi @brianwang12 - PR #8323 has now landed so please could you rebase to resolve the small conflict?

**Comment by brianwang12** [comment]
> Hi @brianwang12 - PR #8323 has now landed so please could you rebase to resolve the small conflict?


> Hi @brianwang12 - PR #8323 has now landed so please could you rebase to resolve the small conflict?

I got it.  I will try to rebase them as soon as possible.

**Comment by brianwang12** [comment]
Update the 'github_get_release_file_url()' function to make it compatible with different architecture expressions. eg. aarch64/arm64, or x86_64/amd64, allowing it to acquire the correct URL addresses.

I tested this patch with Arm and X86 environment.  You can try it out as follows:
  ./kata-manager.sh  -f -k 3.2.0-alpha3 -c v1.7.2
  ./kata-manager.sh  -f -k 3.3.0-alpha0 -c v1.7.2

File: utils/kata-manager.sh
Comment (jodh-intel) [inline utils/kata-manager.sh:185]: I'd prefer that rather than using multiple regexes (which looks odd/wrong in the cases where we don't set `regex2`, we do something like this instead:

```bash
local arch
arch=$(uname -m)

# List of architecture names for the current system that we will look for.
local arches=()

arches+=("$arch")

[ "$arch" = "x86_64" ] && arches+=("amd64")
# XXX: Do something similar for the other arches...

# What follows is architecture agnostic...

local arch_regex
arch_regex=$(echo "${arches[@]}"|tr ' ' '|'|sed 's!|$!!g')
arch_regex=$(printf "(%s)" "$arch_regex")

regex="kata-static-${version}-${arches_regex}.tar.xz"

jq ... | grep -E "/${regex}$"
```

**Review by jodh-intel — COMMENTED** [review]
Thanks @brianwang12.

File: utils/kata-manager.sh
Comment (brianwang12) [reply utils/kata-manager.sh]: Done! Thanks for your detailed guidance and patience. I've followed your instructions to make the code modifications."

**Comment by BbolroC** [comment]
The installation has been successfully done for s390x on Ubuntu 20.04/22.04 with:

```
./kata-manager.sh -f -k 3.2.0-alpha3 -c v1.7.2
./kata-manager.sh -f -k 3.3.0-alpha0 -c v1.7.2
```

File: utils/kata-manager.sh
Comment (jodh-intel) [inline utils/kata-manager.sh:189]: Hmm. I think that as one extra sanity check we should ensure that only 1 match is returned here:

```bash
local lines
lines=$(echo "$download_url" | wc -l | awk '{print $1}')

[ "$lines" -eq 1 ] || die "expected 1 download URL but found '$download_url'"
```

However, since I've already approved it, I'll let others comment on whether we should do that on this PR or whether we could do it on a follow-up PR though.

**Review by jodh-intel — APPROVED** [review]
Thanks @brianwang12. I've re-tested this on x86_64 by running the script as:

```bash
utils/kata-manager.sh -d -f -o
```

**Review by fidencio — APPROVED** [review]
lgtm, and @jodh-intel suggestion can come on a follow-up PR.
Thanks @brianwang12!

File: utils/kata-manager.sh
Comment (jodh-intel) [reply utils/kata-manager.sh:189]: I've raised #8364 to add this check later.

Comment (brianwang12) [reply utils/kata-manager.sh:189]: Thank you for your guidance. It's valuable to include an extra sanity check for improved robustness.

**Pull request merged**
